File: backend/app/schemas/note.py
"""Note-related Pydantic schemas.

Defines request and response models for note CRUD, pinning,
archiving, and tag attachment operations.
"""

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.schemas.tag import TagResponse


class NoteCreate(BaseModel):
    """Schema for creating a new note.

    Attributes:
        title: Note title (max 500 characters, defaults to ``"Untitled"``).
        content: Rich-text editor document as a JSON-compatible dict.
        tag_ids: Optional list of tag UUIDs to attach to the new note.
    """

    title: str = Field(default="Untitled", max_length=500)
    content: dict[str, Any] | None = None
    tag_ids: list[str] | None = None


class NoteUpdate(BaseModel):
    """Schema for updating an existing note.

    All fields are optional; only provided fields will be updated.

    Attributes:
        title: New note title (max 500 characters).
        content: New rich-text editor document.
    """

    title: str | None = Field(None, max_length=500)
    content: dict[str, Any] | None = None


class NotePinUpdate(BaseModel):
    """Schema for toggling a note's pinned status.

    Attributes:
        is_pinned: Whether the note should be pinned to the top.
    """

    is_pinned: bool


class NoteArchiveUpdate(BaseModel):
    """Schema for toggling a note's archived status.

    Attributes:
        is_archived: Whether the note should be archived.
    """

    is_archived: bool


class NoteTagAttach(BaseModel):
    """Schema for attaching a tag to a note.

    Attributes:
        tag_id: UUID of the tag to attach.
    """

    tag_id: str


class NoteResponse(BaseModel):
    """Schema for note API responses.

    Attributes:
        id: Unique note identifier.
        notebook_id: The notebook this note belongs to.
        user_id: The user who owns this note.
        title: Note title.
        content: Rich-text editor document (JSONB-compatible dict).
        plain_text: Extracted plain text for search.
        is_pinned: Whether the note is pinned.
        is_archived: Whether the note is archived.
        notebook_name: Name of the parent notebook (populated at query time).
        archived_at: Timestamp when the note was archived, if applicable.
        created_at: Timestamp when the note was created.
        updated_at: Timestamp when the note was last updated.
        tags: List of tags attached to this note.
    """

    id: uuid.UUID
    notebook_id: uuid.UUID
    user_id: uuid.UUID
    title: str
    content: dict[str, Any] | None
    plain_text: str | None
    is_pinned: bool
    is_archived: bool
    # Populated dynamically via annotation, not stored in the database.
    notebook_name: str | None = None
    archived_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    tags: list[TagResponse] = []

    model_config = {"from_attributes": True}

File: frontend/src/types/note.ts
/**
 * @module types/note
 * @description Type definitions for notes — the core content entity that
 * belongs to a notebook and can be tagged, pinned, and archived.
 */

import type { Tag } from './tag'

/** Represents a note returned by the API. */
export interface Note {
  /** Unique identifier (UUID). */
  id: string
  /** ID of the parent notebook. */
  notebook_id: string
  /** ID of the owning user. */
  user_id: string
  /** Note title. */
  title: string
  /** Rich-text content as a JSON object (e.g. TipTap/ProseMirror doc), or `null` if empty. */
  content: Record<string, any> | null
  /** Plain-text representation of the content, used for search indexing. */
  plain_text: string | null
  /** Whether the note is pinned to the top of its notebook. */
  is_pinned: boolean
  /** Whether the note is archived. */
  is_archived: boolean
  /** ISO-8601 timestamp when the note was archived, or `null`. */
  archived_at: string | null
  /** Name of the parent notebook (optional, populated by some endpoints). */
  notebook_name?: string
  /** ISO-8601 timestamp when the note was created. */
  created_at: string
  /** ISO-8601 timestamp when the note was last updated. */
  updated_at: string
  /** Tags attached to the note (optional, populated by some endpoints). */
  tags?: Tag[]
}

/** Payload for creating a new note. */
export interface NoteCreateRequest {
  /** Note title. */
  title: string
  /** Rich-text content as a JSON object (optional at creation). */
  content?: Record<string, any>
  /** IDs of tags to attach immediately (optional). */
  tag_ids?: string[]
}

/** Payload for updating an existing note. All fields are optional. */
export interface NoteUpdateRequest {
  /** New title. */
  title?: string
  /** New rich-text content. */
  content?: Record<string, any>
  /** Toggle pinned status. */
  is_pinned?: boolean
  /** Toggle archived status. */
  is_archived?: boolean
}
